ifacemonitor/update_filter.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) 2020 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ifacemonitor
+
+import (
+	"context"
+	"net"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+
+	timeshim "github.com/projectcalico/felix/time"
+)
+
+const FlapDampingDelay = 100 * time.Millisecond
+
+func NewUpdateFilter(options ...UpdateFilterOp) *UpdateFilter {
+	u := &UpdateFilter{
+		Time: timeshim.NewRealTime(),
+	}
+	for _, op := range options {
+		op(u)
+	}
+	return u
+}
+
+type UpdateFilter struct {
+	Time timeshim.Time
+}
+
+type UpdateFilterOp func(filter *UpdateFilter)
+
+func WithTimeShim(t timeshim.Time) UpdateFilterOp {
+	return func(filter *UpdateFilter) {
+		filter.Time = t
+	}
+}
+
+// FilterUpdates filters out updates that occur when IPs are quickly removed and re-added.
+// Some DHCP clients flap the IP during an IP renewal, for example.
+//
+// Algorithm:
+// * Maintain a queue of link and address updates per interface.
+// * When we see a potential flap (i.e. an IP deletion), defer processing the queue for a while.
+// * If the flap resolves itself (i.e. the IP is added back), suppress the IP deletion.
+func (u *UpdateFilter) FilterUpdates(ctx context.Context,
+	addrOutC chan<- netlink.AddrUpdate, addrInC <-chan netlink.AddrUpdate,
+	linkOutC chan<- netlink.LinkUpdate, linkInC <-chan netlink.LinkUpdate) {
+
+	logrus.Debug("FilterUpdates: starting")
+	var timerC <-chan time.Time
+
+	type timestampedUpd struct {
+		ReadyAt time.Time
+		Update  interface{} // AddrUpdate or LinkUpdate
+	}
+
+	updatesByIfaceIdx := map[int][]timestampedUpd{}
+
+	for {
+		select {
+		case <-ctx.Done():
+			logrus.Info("FilterUpdates: Context expired, stopping")
+			return
+		case linkUpd := <-linkInC:
+			idx := linkUpd.Index
+			if len(updatesByIfaceIdx[int(idx)]) == 0 {
+				// We do see link updates as part of a flap but we've only seen them after the IP is
+				// flapped down.  Avoid delaying unrelated link updates.
+				logrus.Debug("FilterUpdates: link change with empty queue, short circuit.")
+				linkOutC <- linkUpd
+				continue
+			}
+			// Otherwise, we might have a flap in progress, delay the update.
+			updatesByIfaceIdx[int(idx)] = append(updatesByIfaceIdx[int(idx)],
+				timestampedUpd{
+					ReadyAt: u.Time.Now().Add(FlapDampingDelay),
```

Comment:
Technically any update but DEL is ready now. The queue is sorted by arrival and you release when the head is ready. So if the first one is a DEL is will hold the whole queue. But you do not want to held back any unrelated update for longer. So if you have `[addr 1 del] [addr 2 add]` you want the `[addr 2 add]` to go out asap, don't you? So I wonder if you want to add delay only for DELs.

Reply:
Agreed, that'd be an improvement.

Reply:
I think we at least need to delay the iface down message since, when an interface goes down, we get a race with IP removal messages and iface down arriving on different channels.

When an iface goes up it's typically going to have its IPs added immediately after so delaying the iface up message makes it more likely that the iface monitor will get a complete snapshot of the IPs when it does its first read.

Reply:
Hmm, but that's not what the code does; looks like I need to always delay link down.